src/middlewared/middlewared/plugins/mdns.py

```diff
@@ -231,6 +392,24 @@ def __init__(self, *args, **kwargs):
         self.initialized = False
         self.lock = threading.Lock()
 
+    @private
+    async def get_hostname(self):
+        """
+        Return virtual hostname if the server is HA and this is the active controller.
+        Return None if this is the passive controller.
+        In all other cases return the hostname_local.
+        This is to ensure that the correct hostname is used for mdns advertisements.
+        """
+        ngc = await self.middleware.call('network.configuration.config')
+        if not await self.middleware.call('system.is_freenas') and await self.middleware.call('notfier.failover_licensed'):
```

Comment:
If think this check can be removed if we check for the existence of `hostname_virtual` in ngc?

Reply:
Yes. Good point